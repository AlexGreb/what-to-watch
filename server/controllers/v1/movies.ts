import type {ResponseData} from "../../types";
import type {IMovie} from "../../Models/movie";
import type {Request, Response} from "express";
import * as movieService from "../../services/v1/movieService";
import {getFavoriteMovies} from "../../services/v1/favoriteService";
import {getResError, setFavoriteMovies} from "../../utils"
import {validationResult} from "express-validator";

export const getAllMovies = async (req: Request, res: Response): Promise<void> => {
    try {
        const errors = validationResult(req);
        if(!errors.isEmpty()) {
            res.status(404).json({
                status: 'FAILED',
                data: {
                    error: errors,
                }
            })
        }
        const movies: IMovie[] = await movieService.getAllMovies();
        const favoriteMovies = await getFavoriteMovies(Number(req.params.userId));
        const moviesWithFavoriteUser = setFavoriteMovies(favoriteMovies, movies);
        const data: ResponseData<IMovie[]> = {
            status: 'OK',
            data: moviesWithFavoriteUser
        }
        res.json(data);
    } catch(error: unknown) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}

export const getMovie = async (req: Request, res: Response): Promise<void> => {
    const {params} = req;
    if(!params.movieId) return;
    try {
        const movie: IMovie | null = await movieService.getMovie(Number(params.movieId));

        if(movie) {
            res.status(200).json({
                status: 'OK',
                data: movie,
            })
        } else {
            res.status(404).json({
                status: 'FAILED',
                data: {
                    error: "Movie with this movieId wasn`t found",
                }
            })
        }
    } catch(error) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}

export const getPromo = async (req: Request, res: Response): Promise<void> =>{
    try {
        const promo: IMovie | null = await movieService.getPromo();
        res.json({
            status: 'OK',
            data: promo
        });
    } catch(error) {
        const {status, data} = getResError(error);
        res.status(status).json({
            status: 'FAILED',
            data
        });
    }
}
