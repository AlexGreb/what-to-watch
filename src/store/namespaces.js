export const namespaces = {
  APP: `app`,
  MOVIES: `movies`
};
