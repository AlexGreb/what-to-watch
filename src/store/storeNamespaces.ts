export enum Namespaces {
  APP = `app`,
  MOVIES = `movies`,
  USER = `user`
};
