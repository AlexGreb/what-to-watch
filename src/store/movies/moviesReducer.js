import {createSlice} from '@reduxjs/toolkit';
import {moviesList} from '../../mocks/movies.js';
import {createMovie} from '../../adapters/adapters';

const initialState = {
  moviesList: []
};

const moviesSlice = createSlice({
  name: `movies`,
  initialState,
  reducers: {
    setMoviesList: (state, action) => {
      state.moviesList = action.payload;
    }

  }
});

const {setMoviesList} = moviesSlice.actions;

const Operations = {
  getMoviesList: () => (dispatch) => {
    setTimeout(() => {
      dispatch(setMoviesList(moviesList.map(createMovie)));
    }, 1000);
  }
};

export {setMoviesList, Operations, initialState};
export default moviesSlice.reducer;

