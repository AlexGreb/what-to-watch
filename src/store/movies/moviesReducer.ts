import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {createMovie, createReview} from '../../adapters/adapters';
import {AppRoute} from '../../constants/consts.ts';
import {Namespaces} from '../storeNamespaces';
import {Movie, MovieReviews} from '../../types/movie';

type InitialState {
  promoMovie?: Movie | {},
  loadingMovies: boolean,
  loadingMyMovies: boolean,
  reviews: MovieReviews | [],
  sendingReviewStatus: boolean,
  myListMovies: Array<Movie> | [],
  moviesList: Array<Movie> | [],
  isError: boolean
}

const initialState:InitialState = {
  promoMovie: {},
  loadingMovies: false,
  loadingMyMovies: false,
  reviews: [],
  sendingReviewStatus: false,
  myListMovies: [],
  moviesList: [],
  isError: false
};

const Operations = {
  fetchMoviesList: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchMoviesList`,
      async (_, {extra}) => {
        const response = await extra.get(AppRoute.FILMS);
        return response.data.map(createMovie);
      }),
  fetchPromoMovie: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchPromoMovie`,
      async (_, {extra}) => {
        const response = await extra.get(AppRoute.PROMO);
        const a = await response.json()
        return createMovie(response.data);
      }),
  fetchReviews: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchReviews`,
      async (movieId: number, {extra}) => {
        const response = await extra.get(`${AppRoute.REVIEWS}${movieId}`);
        return response.data.map(createReview);
      }
  ),
  fetchComment: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchComment`,
      async (reviewData, {extra}) => {
        const response = await extra.post(`${AppRoute.REVIEWS}${reviewData.movieId}`, reviewData.review);
        return response.data;
      }
  ),
  fetchMyListMovies: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchMyListMovies`,
      async (_, {extra}) => {
        const response = await extra.get(`${AppRoute.FAVORITE}`);
        return response.data.map(createMovie);
      }
  ),
  fetchChangeFavoriteStateMovie: createAsyncThunk(
      `${Namespaces.MOVIES}/fetchChangeFavoriteStateMovie`,
      async (movie: Movie, {dispatch, fulfillWithValue, extra}) => {
        const response = await extra.post(`${AppRoute.FAVORITE}${movie.id}/${movie.isFavorite ? 0 : 1}`);
        dispatch(Operations.fetchMoviesList());
        dispatch(Operations.fetchPromoMovie());
        return fulfillWithValue(response.data);
      }
  )
};

const moviesSlice = createSlice({
  name: Namespaces.MOVIES,
  initialState,
  reducers: {
    setSendingReviewStatus: (state, action) => {
      state.sendingReviewStatus = action.payload;
    },
  },
  extraReducers: {
    [Operations.fetchMoviesList.fulfilled.type]: (state, action) => {
      state.loadingMovies = false;
      state.moviesList = action.payload;
    },
    [Operations.fetchMoviesList.pending.type]: (state) => {
      state.loadingMovies = true;
    },
    [Operations.fetchMoviesList.rejected.type]: (state) => {
      state.isError = true;
    },
    [Operations.fetchPromoMovie.fulfilled.type]: (state, action) => {
      state.promoMovie = action.payload;
    },
    [Operations.fetchPromoMovie.rejected.type]: (state) => {
      state.isError = true;
    },
    [Operations.fetchReviews.fulfilled.type]: (state, action) => {
      state.reviews = action.payload;
    },
    [Operations.fetchComment.fulfilled.type]: (state) => {
      state.sendingReviewStatus = true;
    },
    [Operations.fetchComment.rejected.type]: (state) => {
      state.sendingReviewStatus = false;
    },
    [Operations.fetchMyListMovies.fulfilled.type]: (state, action) => {
      state.loadingMyMovies = false;
      state.myListMovies = action.payload;
    },
    [Operations.fetchMyListMovies.pending.type]: (state) => {
      state.loadingMyMovies = true;
    },
    [Operations.fetchMyListMovies.rejected.type]: (state) => {
      state.loadingMyMovies = false;
      state.isError = true;
    }
  }
});

export const {setSendingReviewStatus} = moviesSlice.actions;
export {
  Operations,
  initialState
};
export default moviesSlice.reducer;

