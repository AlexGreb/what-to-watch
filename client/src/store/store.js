import {configureStore} from '@reduxjs/toolkit';
import appReducer from './app/appReducer';
import moviesReducer from './movies/moviesReducer.js';
import userReducer from './user/userReducer.js';
import {namespaces} from './namespaces';
import {createApi} from '../api.js';


const api = createApi((...args) => store.dispatch(...args));

export const store = configureStore({
  reducer: {
    [namespaces.APP]: appReducer,
    [namespaces.MOVIES]: moviesReducer,
    [namespaces.USER]: userReducer
  },
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware({
      thunk: {
        extraArgument: api
      },
    });
  },
  devTools: true
  // eslint-disable-next-line no-undef
  // devTools: process.env.NODE_ENV !== `production`,
});

// export type State = ReturnType<typeof store.getState>
