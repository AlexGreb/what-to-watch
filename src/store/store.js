import {configureStore} from '@reduxjs/toolkit';
import appReducer from './app/appReducer.js';
import moviesReducer from './movies/moviesReducer.js';
import {namespaces} from './namespaces.js';

export const store = configureStore({
  reducer: {
    [namespaces.APP]: appReducer,
    [namespaces.MOVIES]: moviesReducer
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  devTools: true
  // eslint-disable-next-line no-undef
  // devTools: process.env.NODE_ENV !== `production`,
});
