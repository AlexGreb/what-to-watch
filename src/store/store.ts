import {configureStore} from '@reduxjs/toolkit';
import appReducer from './app/appReducer';
import moviesReducer from './movies/moviesReducer.js';
import userReducer from './user/userReducer.js';
import {Namespaces} from './storeNamespaces';
import {createApi} from '../api.js';


const api = createApi((...args) => store.dispatch(...args));

export const store = configureStore({
  reducer: {
    [Namespaces.APP]: appReducer,
    [Namespaces.MOVIES]: moviesReducer,
    [Namespaces.USER]: userReducer
  },
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware({
      thunk: {
        extraArgument: api
      },
    });
  },
  devTools: true
  // eslint-disable-next-line no-undef
  // devTools: process.env.NODE_ENV !== `production`,
});

export type State = ReturnType<typeof store.getState>