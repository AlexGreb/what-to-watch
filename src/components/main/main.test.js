/**
 * @jest-environment jsdom
 */
import React from 'react';
import {render} from '@testing-library/react';
import Main from './main.jsx';
import {moviesList} from '../../mocks/movies.js';
import {Provider} from 'react-redux';
import {store} from '../../store/store.js';

describe(`Main`, () => {
  window.HTMLMediaElement.prototype.load = () => {};
  window.HTMLMediaElement.prototype.play = () => {};
  window.HTMLMediaElement.prototype.pause = () => {};

  it(`Should render correctly main page`, () => {
    const {asFragment} = render(
        <Provider store={store}>
          <Main
            moviesList={moviesList}
          />
        </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
