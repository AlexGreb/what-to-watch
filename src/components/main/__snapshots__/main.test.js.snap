// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Should render correctly main page 1`] = `
<DocumentFragment>
  <section
    class="movie-card"
  >
    <header
      class="page-header movie-card__head"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="user-block"
      >
        <div
          class="user-block__avatar"
        >
          <img
            alt="User avatar"
            height="63"
            src="img/avatar.jpg"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      class="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/bg-the-grand-budapest-hotel.jpg"
      />
    </div>
    <h1
      class="visually-hidden"
    >
      WTW
    </h1>
    <div
      class="movie-card__wrap"
    >
      <div
        class="movie-card__info"
      >
        <div
          class="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          class="movie-card__desc"
        >
          <h2
            class="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            class="movie-card__meta"
          >
            <span
              class="movie-card__genre"
            >
              Drama
            </span>
            <span
              class="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            class="movie-card__buttons"
          >
            <button
              class="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlink:href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              class="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlink:href="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="page-content"
  >
    <section
      class="catalog"
    >
      <h2
        class="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        class="catalog__genres-list"
      >
        <li
          class="catalog__genres-item catalog__genres-item--active"
        >
          <a
            class="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
      </ul>
      <div
        class="catalog__movies-list"
      />
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link logo__link--light"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
